### PR TITLE
Port MangaKawaii

### PR DESCRIPTION
**Pull request by MikeZeDev**

### CODE REVIEW

File: web/src/engine/websites/MangaKawaii.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
+import { Tags } from '../Tags';
+import icon from './MangaKawaii.webp';
+import { Chapter, DecoratableMangaScraper, Manga, type MangaPlugin } from '../providers/MangaPlugin';
+import * as Common from './decorators/Common';
+import { FetchCSS, FetchRequest, FetchWindowScript } from '../FetchProvider';
+
+const cdn = 'https://cdn.mangakawaii.pics';
+
+const pagesScript = `
+    new Promise((resolve, reject) => {
```
Comment (ronny1982) [inline web/src/engine/websites/MangaKawaii.ts:10]: Can be simplified due to automatic rejection of synchronous errors in promises.
```typescript
new Promise(resolve => {
    resolve(pages.map(page => '${cdn}/uploads/manga/'+ oeuvre_slug +'/chapters_'+applocale+'/'+chapter_slug+ '/'+page.page_image));
});
```

File: web/src/engine/websites/MangaKawaii.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        request.headers.set('X-CSRF-TOKEN', token);
+        const data = await FetchCSS<HTMLAnchorElement>(request, 'ul.manga-list-text li a.alpha-link');
+        return data.map(element => {
+            const bloat = element.querySelector('span');
+            if (bloat) element.removeChild(bloat);
+            return new Manga(this, provider, element.pathname, element.text.trim());
+        });
+    }
+
+    public override async FetchChapters(manga: Manga): Promise<Chapter[]> {
```
Comment (ronny1982) [inline web/src/engine/websites/MangaKawaii.ts:47]: Suggestion to make this method more flexible by providing languages as array and iterate through array `[ 'en', 'fr' ]`

File: web/src/engine/websites/MangaKawaii.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    async getChaptersFromLanguage(manga: Manga, language: string): Promise<Chapter[]> {
+        await this.changeLanguage(language);
+        const uri = new URL(manga.Identifier, this.URI);
+        const request = new FetchRequest(uri.href);
+        const firstChapter = await FetchCSS<HTMLAnchorElement>(request, 'table.table--manga tbody td.table__chapter a');
+        if (typeof firstChapter[0] === 'undefined') {
+            return [];
+        }
+
+        const uri2 = new URL(firstChapter[0].pathname, this.URI);
```
Comment (ronny1982) [inline web/src/engine/websites/MangaKawaii.ts:62]: May re-use `let uri` and `let request`, or provide a more expressive variable name.

File: web/src/engine/websites/MangaKawaii.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+    }
+    async getToken() {
+        const request = new FetchRequest(this.URI.href);
+        const data = await FetchCSS <HTMLMetaElement>(request, 'meta[name="csrf-token"]');
+        return data[0].content;
+    }
+
+    async changeLanguage(language) {
+        await FetchWindowScript(new FetchRequest(this.URI.href), ''); //necessary cause /lang redirect to last viewed manga
```
Comment (ronny1982) [inline web/src/engine/websites/MangaKawaii.ts:77]: This seems redundant, I think language change should work independently from the last visited URL

**Review by ronny1982 — APPROVED** [review]

File: web/src/engine/websites/MangaKawaii.ts (hunk 4 above)
Comment (MikeZeDev) [reply web/src/engine/websites/MangaKawaii.ts:77]: Its here on purposes. Without it, sometimes languages fails to change (and i end up receiving eng chapters after a /lang/fr request and vice versa).

There is something fishy here.

Comment (MikeZeDev) [reply web/src/engine/websites/MangaKawaii.ts:77]: lang is server sided in a session token

File: web/src/engine/websites/MangaKawaii.ts [hunk 5]
```diff
@@ -45,23 +40,24 @@ export default class extends DecoratableMangaScraper {
     }
 
     public override async FetchChapters(manga: Manga): Promise<Chapter[]> {
-        const englishChapters = await this.getChaptersFromLanguage(manga, 'en');
-        const frenchChapters = await this.getChaptersFromLanguage(manga, 'fr');
-        return [...englishChapters, ...frenchChapters];
+        const languages = ['eng', 'fr'];
+        const chapters = [];
+        languages.forEach(async lang => chapters.push(await this.getChaptersFromLanguage(manga,lang)));
```
Comment (ronny1982) [inline web/src/engine/websites/MangaKawaii.ts:45]: Will `.forEach` wait if the inner callback function is `async`?
May better use an ordinary `for ... of ...` loop instead just to be sure

Comment (ronny1982) [reply web/src/engine/websites/MangaKawaii.ts:45]: Just testet, this will not work, line 46 (return) is hit before the async forEach completes

Comment (MikeZeDev) [reply web/src/engine/websites/MangaKawaii.ts:45]: Oops

**Comment by MikeZeDev** [comment]
They activated Cloudflare.

**Pull request merged**